### PR TITLE
Added support for elevation in refresh indicator.

### PR DESCRIPTION
**Pull request by dishankjindal1**
Refresh Indicator Elevation:

* Added elevation to the refresh indicator using `elevation` attribute.
* This change improves the indicator's visibility against various background colors and enhances overall UI consistency.
* If we want to get rid of shadow effect from the indicator itself.

The refresh indicator did not provide the control over depth of the progress indicator. This change addresses that issue by adding elevation to the indicator, creating a subtle depth effect that makes it more visually prominent and provides better feedback to the user.

No breaking changes.

| Before | After |
| --- | --- |
| before| after |

## Pre-launch Checklist

- [yes] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [yes] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [yes] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [yes] I signed the [CLA].
- [yes] I listed at least one issue that this PR fixes in the description above.
- [yes] I updated/added relevant documentation (doc comments with `///`).
- [yes] I added new tests to check the change I am making, or this PR is [test-exempt].
- [yes] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [yes] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/refresh_indicator.dart [hunk 1]
```diff
@@ -591,6 +598,7 @@ class RefreshIndicatorState extends State<RefreshIndicator> with TickerProviderS
                       valueColor: _valueColor,
                       backgroundColor: widget.backgroundColor,
                       strokeWidth: widget.strokeWidth,
+                        elevation: widget.elevation,
```
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/material/refresh_indicator.dart:601]: Formatting nit: some extra space here. `elevation` needs to be lined up with the other proerties.

**Review by MitchellGoodwin — COMMENTED** [review]
Thank you for the PR! This will need a test to protect against future regressions. Just a simple test that checks if the elevation is the expected value both by default and when using a custom value. Also, I'm not sure why the golden test failed before, we'll have to keep an eye on that.

**Comment by Piinks** [comment]
Hey @dishankjindal1 would you like to continue working on this change?

**Comment by dishankjindal1** [comment]
Hey @Piinks - Yes, I will add the tests. Sorry for the delayed response. Was busy working on other stuff.

**Comment by dishankjindal1** [comment]
Hello @Piinks @MitchellGoodwin - Please review my latest PR including the test cases. And minor formatting fix.

File: packages/flutter/lib/src/material/refresh_indicator.dart [hunk 2]
```diff
@@ -143,6 +143,7 @@ class RefreshIndicator extends StatefulWidget {
     this.semanticsValue,
     this.strokeWidth = RefreshProgressIndicator.defaultStrokeWidth,
     this.triggerMode = RefreshIndicatorTriggerMode.onEdge,
+    this.elevation = 0,
```
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/material/refresh_indicator.dart:146]: It looks like `RefreshProgressIndicator` has a default of `2` so this would be a minor breaking change as this would be different default behavior. Was there a reason to make it 0?

Comment (dishankjindal1) [reply packages/flutter/lib/src/material/refresh_indicator.dart:146]: Yes, elevation is missing refresh indicator api up until current flutter sdk. All the current user base will be affected as they might not need elevation. Hence keeping the elevation 0 will not break current ui functionality. 

That is what my thought process is.

@MitchellGoodwin What would you suggest? or this is fine? because i wrote the new test-cases and non of the existing test cases are failing

Comment (MitchellGoodwin) [reply packages/flutter/lib/src/material/refresh_indicator.dart:146]: They did actually have a hardcoded elevation before of 2, because nothing was provided to the `RefreshIndicator` it used it's default value of 2. https://github.com/flutter/flutter/blob/f2ba2049174d28b08d7466946eb4362bbec6b821/packages/flutter/lib/src/material/progress_indicator.dart#L881

So by having this provide a default of `0` we are changing the default elevation from 2 to 0. It also looks like the Material 2 default refresh indicator should have some elevation:
https://m2.material.io/design/platform-guidance/android-swipe-to-refresh.html#usage

I think it would be best to have a default of 2, and adjust the test accordingly.

**Comment by dishankjindal1** [comment]
@MitchellGoodwin - Please checkout my latest commit, and review my PR again. Everything seems in correct order now.

**Review by MitchellGoodwin — APPROVED** [review]
LGTM! Thank you for the contribution!

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 3]
```diff
@@ -1203,4 +1203,107 @@ void main() {
       await testColor(color);
     }
   });
+    testWidgets('RefreshIndicator default elevation value correctly',
+      (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/material/refresh_indicator_test.dart:1207]: It looks like this should be outdented by 2 spaces.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 4]
```diff
@@ -1203,4 +1203,107 @@ void main() {
...
+        expect(
+          getIndicator().elevation,
+          2.0,
+        );
+        // Wait until the fling finishes before starting the next fling.
+        await tester.pumpAndSettle();
+      }
+    }
+
+    testElevation();
```
Comment (justinmc) [inline packages/flutter/test/material/refresh_indicator_test.dart:1249]: I think you can get rid of this function since it's only called once.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 5]
```diff
@@ -1203,4 +1203,107 @@ void main() {
...
+          elevation,
+        );
+        // Wait until the fling finishes before starting the next fling.
+        await tester.pumpAndSettle();
+      }
+    }
+
+    for (final double elevation in elevations) {
+      await testElevation(elevation);
+    }
```
Comment (justinmc) [inline packages/flutter/test/material/refresh_indicator_test.dart:1306]: Here's how I would typically do this kind of thing:

```dart
for (final double elevation in <double>[0, 0.1, 1, 1.1, 1.59]) {
  testWidgets('RefreshIndicator uses elevation value $elevation correctly',
      (WidgetTester tester) async {
    // ... all of your test code.
  });
}
```

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 6]
```diff
@@ -1203,4 +1203,107 @@ void main() {
...
+  });
+
```
Comment (justinmc) [inline packages/flutter/test/material/refresh_indicator_test.dart:1308]: No empty line needed here.

**Review by justinmc — COMMENTED** [review]
Some quick formatting things, otherwise this looks great, thanks!

File: packages/flutter/lib/src/material/refresh_indicator.dart [hunk 7]
```diff
@@ -143,6 +143,7 @@ class RefreshIndicator extends StatefulWidget {
     this.semanticsValue,
     this.strokeWidth = RefreshProgressIndicator.defaultStrokeWidth,
     this.triggerMode = RefreshIndicatorTriggerMode.onEdge,
+    this.elevation = 2,
```
Comment (victorsanni) [inline packages/flutter/lib/src/material/refresh_indicator.dart:146]: Really really tiny nit: Since the parameter is a double, it should have a default of 2.0 instead of 2.

Comment (dishankjindal1) [reply packages/flutter/lib/src/material/refresh_indicator.dart]: @MitchellGoodwin - Not a problem. This is tiny but a very valid observation. Thanks for noticing. Pushed a new commit.

File: packages/flutter/test/material/refresh_indicator_test.dart (hunk 5 above)
Comment (dishankjindal1) [reply packages/flutter/test/material/refresh_indicator_test.dart:1306]: @justinmc I took reference from the existing test cases written above my code.

File: packages/flutter/test/material/refresh_indicator_test.dart (hunk 6 above)
Comment (dishankjindal1) [reply packages/flutter/test/material/refresh_indicator_test.dart:1308]: @justinmc empty lines also also present in other *_test.dart files.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 8]
```diff
@@ -1203,4 +1203,102 @@ void main() {
       await testColor(color);
     }
   });
+  testWidgets('RefreshIndicator default elevation value correctly',
+    (WidgetTester tester) async {
+  const List<double> positions = <double>[50.0, 100.0, 150.0];
+
+    final AnimationController positionController =
+        AnimationController(vsync: const TestVSync());
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1211]:
```suggestion
    final AnimationController positionController = AnimationController(vsync: const TestVSync());
```
https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#prefer-a-maximum-line-length-of-80-characters

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 9]
```diff
@@ -1203,4 +1203,102 @@ void main() {
       await testColor(color);
     }
   });
+  testWidgets('RefreshIndicator default elevation value correctly',
+    (WidgetTester tester) async {
+  const List<double> positions = <double>[50.0, 100.0, 150.0];
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1208]:
```suggestion
    const List<double> positions = <double>[50.0, 100.0, 150.0];
```

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 10]
```diff
@@ -1203,4 +1203,102 @@ void main() {
...
+          onRefresh: refresh,
+          child: ListView(
+            physics: const AlwaysScrollableScrollPhysics(),
+            children: const <Widget>[Text('X')],
+          ),
+        ),
+      ),
+    );
+
+    RefreshProgressIndicator getIndicator() {
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1226]: I would suggest avoiding defining functions as much as possible, especially in tests. The more layers a reader has to get through to figure out what's going on within a block of code, the more time is wasted. Especially for functions that don't encapsulate much, or are only used a few amount of times.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 11]
```diff
@@ -1203,4 +1203,102 @@ void main() {
...
+      return tester.widget<RefreshProgressIndicator>(
+        find.byType(RefreshProgressIndicator),
+      );
+    }
+
+    // Correspond to [_kDragContainerExtentPercentage].
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1232]:
```suggestion
    // Corresponds to [_kDragContainerExtentPercentage].
```

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 12]
```diff
@@ -1203,4 +1203,102 @@ void main() {
...
+    final double maxPosition =
+        tester.view.physicalSize.height / tester.view.devicePixelRatio * 0.25;
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1234]:
```suggestion
    final double maxPosition = tester.view.physicalSize.height / tester.view.devicePixelRatio * 0.25;
```

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 13]
```diff
@@ -1203,4 +1203,102 @@ void main() {
...
+      0,
+      0.1,
+      1,
+      1.1,
+      1.59,
+    ];
+
+    const List<double> positions = <double>[50.0, 100.0, 150.0];
+
+    Future<void> testElevation(double elevation) async {
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1260]: The entire `testWidgets` function is meant to test this exact behavior, so, as before, this function is not needed.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 14]
```diff
@@ -1203,4 +1203,102 @@ void main() {
...
+
+      RefreshProgressIndicator getIndicator() {
+        return tester.widget<RefreshProgressIndicator>(
+          find.byType(RefreshProgressIndicator),
+        );
+      }
+
+      // Correspond to [_kDragContainerExtentPercentage].
+      final double maxPosition =
+          tester.view.physicalSize.height / tester.view.devicePixelRatio * 0.25;
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1286]:
```suggestion
      final double maxPosition = tester.view.physicalSize.height / tester.view.devicePixelRatio * 0.25;
```

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 15]
```diff
@@ -1203,4 +1203,102 @@ void main() {
...
+            onRefresh: refresh,
+            child: ListView(
+              physics: const AlwaysScrollableScrollPhysics(),
+              children: const <Widget>[Text('X')],
+            ),
+          ),
+        ),
+      );
+
+      RefreshProgressIndicator getIndicator() {
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1278]: As before, this function might be unneeded, serving to complicate instead of make easier.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 16]
```diff
@@ -1203,4 +1203,102 @@ void main() {
...
+        return tester.widget<RefreshProgressIndicator>(
+          find.byType(RefreshProgressIndicator),
+        );
+      }
+
+      // Correspond to [_kDragContainerExtentPercentage].
```
Comment (victorsanni) [inline packages/flutter/test/material/refresh_indicator_test.dart:1284]:
```suggestion
      // Corresponds to [_kDragContainerExtentPercentage].
```

File: packages/flutter/test/material/refresh_indicator_test.dart (hunk 15 above)
Comment (dishankjindal1) [reply packages/flutter/test/material/refresh_indicator_test.dart:1278]: @victorsanni - Please check the above few test-cases which are already part of the master branch. On the other hand, I understand these are subjective and varies person to person. I feel perfectly fine reading the function because the naming is clear as the name suggest what it will resolve into. If there is a clear guide on writing a test case please share the link.

Comment (victorsanni) [reply packages/flutter/test/material/refresh_indicator_test.dart:1262]: > I understand these are subjective and varies person to person.
I do agree that these are subjective. I would look at it from the perspective of someone looking at the code for the first time. They would do something like this:

1. Get to `tester.fling`. The understanding of the viewer here is that a refresh is triggered.
2. Get to the `expect` block. Now, they would see the `getIndicator` function. While they would (most likely) guess that the tester is finding the `RefreshProgressIndicator` created in the test, there are also other possible guesses. Maybe the `getIndicator` function is calling some global indicator created somewhere else, possibly outside the scope of this specific test. Maybe its getting another widget class underneath RefreshProgressIndicator (a render object, for instance) instead of an instance of a `RefreshProgressIndicator`.

This ambiguity leads to more time wasted trying to figure out what's happening. We should strive to eliminate ambiguity whenever possible. 

> I feel perfectly fine reading the function because the naming is clear as the name suggest what it will resolve into.

We shouldn't think that because something seems clear to me, or you, it would also seem clear to others. Usually, that's not the case. In fact, something that seems clear to me right now might not seem clear to me after some time.

> Please check the above few test-cases which are already part of the master branch.
We shouldn't keep doing something that might not be right because that's how it's always been done. That leads to mistakes and bad practices multiplying and persisting. Instead, we should strive to do what's best in any given scenario.

I understand this specific use case seems trivial, but small positive changes add up over time to create a product/system that everyone that comes across loves. :)

Comment (dishankjindal1) [reply packages/flutter/test/material/refresh_indicator_test.dart:1262]: > 2. Get to the `expect` block. Now, they would see the `getIndicator` function. While they would (most likely) guess that the tester is finding the `RefreshProgressIndicator` created in the test, there are also other possible guesses. Maybe the `getIndicator` function is calling some global indicator created somewhere else, possibly outside the scope of this specific test. Maybe its getting another widget class underneath RefreshProgressIndicator (a render object, for instance) instead of an instance of a `RefreshProgressIndicator`.
@victorsanni - Honestly, I feel like you are stretching this a lot. This "getIndicator" function is just few lines above the code. If a person who is looking at the code findings it difficult to locate. I am afraid no one can help.

> We shouldn't keep doing something that might not be right because that's how it's always been done. That leads to mistakes and bad practices multiplying and persisting. Instead, we should strive to do what's best in any given scenario.
@victorsanni Are you suggesting, other reviewers were not good? That they were not able to point of the obvious issues. Which is an issue according to you! 

I already told you this approach is fine by me. Unless and until, I am not going against the code structure guidelines. 

Peace✌🏼

Comment (victorsanni) [reply packages/flutter/test/material/refresh_indicator_test.dart:1259]: @dishankjindal1 when and where to encapsulate code into functions is something every engineer has a personal style/preference for.

As you said, there isn't any formal rule in the style guide as to this, and so in no way am I trying to block your PR from being landed. I only hope that the reasons I listed helped clarify the pros and cons of having that function in this situation.

> Are you suggesting, other reviewers were not good? That they were not able to point of the obvious issues. Which is an issue according to you!

It is my responsibility as a reviewer to raise concerns I might have with regards to changes made in a PR. This is independent of the opinions of other reviewers. If I see something that I think doesn't look good, or could be better, I will articulate why I hold that view. As I have already said, I only hope the reasons I gave made sense to you, and to anyone else following the conversation. 

In the end, as you've mentioned, it is a trivial issue, and I want to reiterate that I'm not intentionally trying to block your PR from being landed. I'm only here to help, and I hope we can both learn and grow from this experience to be better engineers. Thank you!

Comment (dishankjindal1) [reply packages/flutter/test/material/refresh_indicator_test.dart]: @victorsanni - I made the changes as requested. Everything looks good to me now.

Comment (victorsanni) [reply packages/flutter/test/material/refresh_indicator_test.dart]: It does feel a little long. If you prefer, you can define the finder a line above to make it a little more readable. But all this is personal preference too. I should have LGTMed even before this conversation, my mistake.

**Review by victorsanni — APPROVED** [review]

**Comment by dishankjindal1** [comment]
Hi @justinmc, Can you go through the Pr once? And do the needful.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 17]
```diff
@@ -1141,16 +1141,7 @@ void main() {
   });
```
Comment (Piinks) [inline packages/flutter/test/material/refresh_indicator_test.dart:1141]: @dishankjindal1 can you check the indentation from earlier comments? Some looks like 4 spaces, the flutter style guide prefers two. That may be what remains to resolve here.

File: packages/flutter/test/material/refresh_indicator_test.dart (hunk 11 above)
Comment (justinmc) [reply packages/flutter/test/material/refresh_indicator_test.dart:1204]: This shouldn't be resolved, it still says "Correspond". Also you can probably remove the square brackets since they don't do anything in regular comments, only `///` doc comments.

File: packages/flutter/test/material/refresh_indicator_test.dart (hunk 16 above)
Comment (justinmc) [reply packages/flutter/test/material/refresh_indicator_test.dart:1234]: +1, this is still not fixed.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 18]
```diff
@@ -1178,29 +1169,78 @@ void main() {
...
+            physics: const AlwaysScrollableScrollPhysics(),
+            children: const <Widget>[Text('X')],
+          ),
+        ),
+      ),
+    );
+
+    // Correspond to [_kDragContainerExtentPercentage].
+    final double maxPosition = tester.view.physicalSize.height / tester.view.devicePixelRatio * 0.25;
+    for (final double position in <double>[50.0, 100.0, 150.0]) {
```
Comment (justinmc) [inline packages/flutter/test/material/refresh_indicator_test.dart:1206]: Is all this flinging really necessary, or can you just do the expect once and be done?

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 19]
```diff
@@ -1178,29 +1169,78 @@ void main() {
...
+          tester.widget<RefreshProgressIndicator>(find.byType(RefreshProgressIndicator)).valueColor!.value!.alpha,
           valueColorAnimation.value!.alpha,
         );
         // Wait until the fling finishes before starting the next fling.
         await tester.pumpAndSettle();
       }
     }
+  });
+
+  testWidgets('RefreshIndicator default elevation value correctly', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/material/refresh_indicator_test.dart:1188]: Nit: "RefreshIndicator passes the default elevation through correctly"

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 20]
```diff
@@ -1178,29 +1169,78 @@ void main() {
...
+              physics: const AlwaysScrollableScrollPhysics(),
+              children: const <Widget>[Text('X')],
+            ),
+          ),
+        ),
+      );
+
+      // Correspond to [_kDragContainerExtentPercentage].
+      final double maxPosition = tester.view.physicalSize.height / tester.view.devicePixelRatio * 0.25;
+      for (final double position in <double>[50.0, 100.0, 150.0]) {
```
Comment (justinmc) [inline packages/flutter/test/material/refresh_indicator_test.dart:1236]: Same comment about this being unnecessary.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 21]
```diff
@@ -1178,29 +1169,78 @@ void main() {
...
+      expect(tester.widget<RefreshProgressIndicator>(find.byType(RefreshProgressIndicator)).elevation, 2.0);
+      // Wait until the fling finishes before starting the next fling.
+      await tester.pumpAndSettle();
+    }
+  });
 
-    for (final Color color in colors) {
-      await testColor(color);
+  testWidgets('RefreshIndicator custom elevation values correctly', (WidgetTester tester) async {
+    for (final double elevation in <double>[0, 0.1, 1, 1.1, 1.59]) {
```
Comment (justinmc) [inline packages/flutter/test/material/refresh_indicator_test.dart:1217]: Maybe just try `0.0`, `2.0`, and one other positive value? Or is there a reason to test all of these?

**Review by justinmc — COMMENTED** [review]
Possibly some unnecessary work in the tests, otherwise small comments here. Sorry for the delay.

**Comment by justinmc** [comment]
Also heads up that there is a merge conflict.

File: packages/flutter/test/material/refresh_indicator_test.dart (hunk 21 above)
Comment (dishankjindal1) [reply packages/flutter/test/material/refresh_indicator_test.dart:1217]: Just for variety i added these many. But i think 0.0 and 2.0 should be enough.

@justinmc - Also, I did not write a test-case for negative elevation values. Let me add these assertion and a unit test to fortify those use-cases.

cc: @victorsanni

File: packages/flutter/test/material/refresh_indicator_test.dart (hunk 18 above)
Comment (dishankjindal1) [reply packages/flutter/test/material/refresh_indicator_test.dart:1206]: Not necessary, but again i am referring test-cases from above. 

I believe one will suffice as well.

**Comment by MitchellGoodwin** [comment]
@dishankjindal1 I see you opened another PR at #154182 and closed this one. Should we switch to reviewing that new one? Was closing this one intentional?

**Comment by dishankjindal1** [comment]
> @dishankjindal1 I see you opened another PR at #154182 and closed this one. Should we switch to reviewing that new one? Was closing this one intentional?

Yes @MitchellGoodwin.

cc: @Piinks @justinmc @victorsanni
